File: tools/check-custody.mjs
#!/usr/bin/env node
// Reference checker for the language-neutral NDF custody-chain vectors.

import { readFileSync } from "node:fs";
import { createHash } from "node:crypto";

function canonicalize(value) {
  if (value === null || typeof value === "boolean" || typeof value === "number" || typeof value === "string") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return "[" + value.map(canonicalize).join(",") + "]";
  return "{" + Object.keys(value).sort().map(
    key => JSON.stringify(key) + ":" + canonicalize(value[key])
  ).join(",") + "}";
}

function validate(events) {
  const errors = [];
  let previous = null;
  let ndfId = null;
  events.forEach((event, index) => {
    if (event.sequence !== index) errors.push("non-contiguous sequence");
    if (event.previous_event_hash !== previous) errors.push("broken previous hash");
    if (ndfId !== null && event.ndf_id !== ndfId) errors.push("ndf_id changed");
    ndfId = event.ndf_id;
    const unsigned = { ...event };
    delete unsigned.event_hash;
    const actual = "sha256:" + createHash("sha256").update(canonicalize(unsigned)).digest("hex");
    if (actual !== event.event_hash) errors.push("incorrect event hash");
    previous = event.event_hash;
  });
  return errors;
}

const file = process.argv[2];
if (!file) throw new Error("usage: node tools/check-custody.mjs <events.json>");
const errors = validate(JSON.parse(readFileSync(file, "utf8")));
for (const error of errors) console.error(error);
process.exitCode = errors.length ? 1 : 0;

File: tools/check_custody.py
#!/usr/bin/env python3
"""Valida uma cadeia de custódia NDF (implementação Python de referência)."""

import argparse
import hashlib
import json
from pathlib import Path

import rfc8785
from jsonschema import Draft202012Validator, FormatChecker


ROOT = Path(__file__).parent.parent


def validate(events):
    schema = json.loads((ROOT / "specs/ndf/schemas/custody-event.schema.json").read_text())
    validator = Draft202012Validator(schema, format_checker=FormatChecker())
    errors = []
    previous = None
    ndf_id = None
    for index, event in enumerate(events):
        errors.extend(f"evento {index}: {e.message}" for e in validator.iter_errors(event))
        if event.get("sequence") != index:
            errors.append(f"evento {index}: sequence não contígua")
        if event.get("previous_event_hash") != previous:
            errors.append(f"evento {index}: previous_event_hash incorrecto")
        if ndf_id is not None and event.get("ndf_id") != ndf_id:
            errors.append(f"evento {index}: ndf_id mudou dentro da cadeia")
        ndf_id = event.get("ndf_id")
        unsigned = {k: v for k, v in event.items() if k != "event_hash"}
        actual = "sha256:" + hashlib.sha256(rfc8785.dumps(unsigned)).hexdigest()
        if event.get("event_hash") != actual:
            errors.append(f"evento {index}: event_hash incorrecto")
        previous = event.get("event_hash")
    return errors


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("file", type=Path)
    args = parser.parse_args()
    events = json.loads(args.file.read_text(encoding="utf-8"))
    errors = validate(events)
    for error in errors:
        print(error)
    return 1 if errors else 0


if __name__ == "__main__":
    raise SystemExit(main())
